scio-core/src/main/scala/com/spotify/scio/runners/dataflow/DataflowResult.scala

```diff
@@ -0,0 +1,124 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.runners.dataflow
+
+import java.util.Collections
+
+import com.google.api.services.dataflow.model.{Job, JobMetrics}
+import com.spotify.scio.metrics.Metrics
+import com.spotify.scio.{RunnerResult, ScioResult}
+import org.apache.beam.runners.dataflow.options.DataflowPipelineOptions
+import org.apache.beam.runners.dataflow.{DataflowClient, DataflowPipelineJob}
+import org.apache.beam.sdk.options.PipelineOptionsFactory
+import org.apache.beam.sdk.runners.AppliedPTransform
+import org.apache.beam.sdk.transforms.PTransform
+import org.apache.beam.sdk.values.{PInput, POutput}
+import org.apache.beam.sdk.{Pipeline, PipelineResult}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+
+/** Represent a Dataflow runner specific result. */
+class DataflowResult(val internal: DataflowPipelineJob) extends RunnerResult {
+
+  def this(internal: PipelineResult) = this(internal.asInstanceOf[DataflowPipelineJob])
+
+  private val client = DataflowResult.getOptions(internal.getProjectId).getDataflowClient
+
+  /** Get Dataflow [[Job]]. */
+  lazy val getJob: Job = client.projects().jobs()
+    .get(internal.getProjectId, internal.getJobId)
+    .setView("JOB_VIEW_ALL")
+    .execute()
+
+  /** Get Dataflow [[JobMetrics]]. */
+  lazy val getJobMetrics: JobMetrics = client.projects().jobs()
+    .getMetrics(internal.getProjectId, internal.getJobId)
+    .execute()
+
+  /** Get a generic [[ScioResult]]. */
+  override def asScioResult: ScioResult = new DataflowScioResult(internal)
+
+  private class DataflowScioResult(internal: PipelineResult) extends ScioResult(internal) {
+    override val finalState: Future[PipelineResult.State] = {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      Future {
+        internal.waitUntilFinish()
+        this.state
+      }
+    }
+
+    override def getMetrics: Metrics = {
+      val options = getJob.getEnvironment.getSdkPipelineOptions.get("options")
+        .asInstanceOf[java.util.Map[String, AnyRef]]
+      Metrics(
+        options.get("scioVersion").toString,
+        options.get("scalaVersion").toString,
+        options.get("appName").toString,
+        internal.getState.toString,
+        getBeamMetrics)
+    }
+  }
+
+}
+
+/** Companion object for [[DataflowResult]]. */
+object DataflowResult {
+  /** Create a new [[DataflowResult]] instance. */
+  def apply(projectId: String, jobId: String): DataflowResult = {
+    val options = getOptions(projectId)
+
+    val job = options.getDataflowClient.projects().jobs()
+      .get(options.getProject, jobId)
+      .setView("JOB_VIEW_ALL")
+      .execute()
+    val transformStepNames: Map[AppliedPTransform[_, _, _], String] =
```

Comment:
Where did you get this from?

Reply:
Trial and error, a lot of :D

Reply:
Maybe add a comment here that explains why this exists, and what you are doing here?